Feature/#9 LoginView + AppleLogin(Sign In with Apple) 구현

minsson opened the pull request:
@jcrescent61 

안녕하세요, 엘렌!
LoginView 화면 및 AppleLogin(Sign In with Apple) 구현을 마무리하여 PR 드립니다 ☺️

## 화면 기획 설명
[LoginView - Figma 링크](https://www.figma.com/file/sJivKiwL6gizBvXWHGEtBH/galpi-Ver.1.0.0?node-id=1%3A2&t=irGNdlz1TnOaoBbl-1)
- 앱을 처음 켰을 때 나오는 로그인 화면입니다.
- 현재는 기획상 애플 로그인만 지원합니다.
- 애플 로그인 버튼을 눌러 로그인을 완료하면 다음 화면으로 넘어갑니다.
- 한 번 로그인하면 앱을 종료한 후 재실행해도 자동으로 로그인되어 다시 로그인 버튼을 누를 필요가 없습니다.

## 구현 내용
- 화면 기획 설명에 해당하는 내용을 구현했습니다.
- Apple의 Apple Login(Sign In with Apple)을 구현했습니다.
- Firebase의 인증 모듈인 FirebaseAuth를 통해 Firebase 로그인을 구현했습니다.
- 전체적인 로그인 과정을 간략하게 다루자면 아래와 같습니다.
    - AppleLogin을 담당하는 객체가 ASAuthorizationController를 통해 인증 Request를 보냅니다.
    - 성공하면 ASAuthorizationControllerDelegate를 통해 ASAuthoriation을 받습니다.
    - ASAuthorization의 credential에서 FirebaseAuth에 필요한 정보를 추출합니다.
    - FirebaseAuth를 담당하는 객체에게 필요한 정보를 넘겨줍니다.
    - FirebaseAuth를 담당하는 객체가 FirebaseAuth를 수행합니다.
    - FirebaseAuth를 통한 인증이 완료되면 다음 화면(TabBarContainerView)으로 넘어갑니다.
- 위 과정에서 여러 객체로 추상화하는 과정이 있었습니다.
    - Login
        - `LoginView`: 애플 로그인 버튼이 있는 로그인 뷰 자체입니다.
        - `LoginViewModel`: 로그인 뷰로부터 로그인 버튼이 탭 됐음을 전달 받아 ServerLoginManagerProtocol 타입에 인증(로그인)하라는 메시지를 보냅니다.
    - OAuth
        - `ServerLoginManagerProtocol`: FirebaseLoginManager를 추상화한 프로토콜입니다.
        - `FirebaseLoginManager`: FirebaseAuth를 통한 인증을 책임지는 객체입니다. SocialLoginManagerProtocol 타입 객체를 통해 소셜 인증 후, 해당 정보를 바탕으로 FirebaseAuth를 진행 후 성공 여부를 LoginViewModel에 알려줍니다.
        - `SocialLoginCredential`: FirebaseAuth에 필요한 값들을 모아놓기 위한 Model 객체입니다.
        - `AuthSecurityService`: randomNonceString과 SHA256 암호화를 진행하는 역할의 객체입니다. (nonce를 SHA256으로 암호화해 Sign In with Apple의 Request를 보낼 때 request.nonce에 할당합니다.
    - SocialLogin
        - `SocialLoginManagerProtocol`: AppleLoginManager를 추상화한 프로토콜입니다. 추후 어떤 소셜로그인 기능이 추가되어도 이 타입을 채택하면 됩니다.
        - `AppleLoginManager`: SocialLoginManagerProtocol을 채택한 객체입니다. ServerLoginManagerProtocol 타입으로부터 인증하라는 메시지를 받아 AppleLogin을 처리하고, 완료 여부를 알려줍니다.
        - `SocialLogin`: 추후 어떤 소셜로그인 기능이 추가되면 해당 객체 정의 후 여기에 case를 추가 후, loginManager 연산 프로퍼티에서 해당 객체를 이니셜라이즈하여 반환하면 됩니다.


## 고민했던 점
- [SignInWithAppleButton](https://developer.apple.com/documentation/authenticationservices/signinwithapplebutton) vs [ASAuthorizationController](https://developer.apple.com/documentation/authenticationservices/asauthorizationcontroller) 
    - 처음에는 너무나 간편한 SignInWithAppleButton API를 사용했습니다. 구현을 완료한 후 결국 ASAuthorizationController API로 모두 리팩토링했습니다. 이유는 아래와 같습니다.
        - SignInWithAppleButton API: 구현이 정말 간단합니다. 코드 몇 줄이면 끝납니다. 대신 LoginView에서 Request 내용도 정의해주고, 로그인이 완료되었을 때 수행할 로직도 작성해줘야 합니다. 즉, 'UI에 대한 책임만 진다'는 게 불가능해집니다.
        - ASAuthorizationController: 구현이 훨씬 복잡합니다. 대신 구체적인 역할별로 객체를 정의해 책임을 더 정확하게 분배할 수 있습니다. 각 객체의 책임과 역할에 좀 더 집중하고 싶어서 이 방안을 선택했습니다.
- MVVM 자체가 적절한지
    - MVVM을 위한 MVVM을 가급적이면 피하고 싶었습니다.
    - 처음에는 MV(Model-View) 패턴으로도 충분한 게 아닌가 고민이었지만, 뷰는 단순해보일지라도 버튼이 눌린 후의 로직이 생각보다 복잡해 MV로 하기에는 각 객체의 역할이 과도하게 무거워지고, 책임이 모호해졌습니다.
        - 특히나 SignInWithAppleButton API를 사용하지 않고 기존 API를 사용하여 코드가 더욱 복잡해졌습니다.
    - 따라서 MVVM을 쓸 이유가 충분하다는 생각이 들었습니다.
    - 그 외에도 계속 고민하며 리팩토링하는 과정이 많았지만, 너무 세세한 내용이므로 생략하겠습니다.
- Sign In with Apple의 Request 내용
    - 우선 fullname과 email을 요청하도록 해두었습니다.
    - 실제 앱 배포시에는 두 정보 모두 요청하지 않는 게 나을 것 같다고 생각합니다.
        - 현재 단계에서 필수적으로 수집해야할 정보가 아니기 때문에, 심사시 거절 당할 가능성이 높아보이기 때문입니다.


## 구현 화면
| 앱 설치 후 최초 로그인  | 로그아웃 후 다시 로그인 |
| :---: | :---: |
| <video src="https://user-images.githubusercontent.com/96630194/220939792-6bd2b020-7bfd-4596-9db2-3a3aa88a1ce3.mp4">  |  <video src="https://user-images.githubusercontent.com/96630194/220943060-431daae5-a21a-4f9a-907f-1cf255fd79aa.mp4"> |

- 현재 로그아웃 기능은 PR에 포함되어 있지 않습니다. 정상 작동 여부를 확인하기 위해 임시로 구현해 넣었으며, 추후 MyPageView를 구현하면서 로그아웃 기능을 추가할 예정입니다.

## 개선하고 싶은 점
- LoginView가 열려있는 상태에서 Light&Dark 모드를 변경시, 모드 변경이 정상 적용되지만 SignInWithAppleButton의 색은 변경되지 않습니다. 대신 모드를 변경한 상태에서 앱을 종료 후 재실행하면 SignInWithAppleButton의 색 또한 정상적으로 변경됩니다. 유저가 이런 일을 겪을 가능성은 매우 희박해보이지만, 나중에라도 해결하고 싶습니다. 

| LoginView 표시 중 Light&Dark 모드 변경  | 앱 종료 후 재실행 |
| :---: | :---: |
| <video src="https://user-images.githubusercontent.com/96630194/220968714-07dad6c9-7074-45d0-8c8f-232658a89f6b.mp4">  |  <video src="https://user-images.githubusercontent.com/96630194/220970946-16dea3cf-41db-427d-989a-e6a9bbe47671.mp4"> |

### linked issue
- #9

minsson committed:
#9 chore: GoogleService-Info 파일 추가 및 gitignore로 관리

- 실제 출시할 앱이므로 GoogleService-Info.plist 파일은 프로젝트에 추가함과 동시에 gitignore에 추가했습니다.

minsson committed:
#9 feat: SPM으로 Firebase SDK 추가

- FirebaseAnalytics, FirebaseAnalyticsSwift, FirebaseAuth를 추가했습니다.

minsson committed:
#9 feat: Firebase Initialization 코드 추가

- Firebase 공식문서에서 안내하는 내용을 가져와 입력했습니다.

minsson committed:
#9 feat: LoginView 파일 생성

- Sign in with Apple 버튼이 위치할 화면입니다.
- 추후 다른 소셜 로그인 기능이 추가될 수 있습니다.

minsson committed:
#9 feat: AuthenticationViewModel 파일 생성

- LoginView에서 사용할 뷰모델입니다.

minsson committed:
#9 feat: SignInWithAppleButton UI 구현

- iOS 14부터 사용 가능한 SignInWithAppleButton의 이니셜라이저를 사용했습니다.

SignInWithAppleButton.init(
    _ label: SignInWithAppleButton.Label = .signIn,
    onRequest: @escaping (ASAuthorizationAppleIDRequest) -> Void,
    onCompletion: @escaping ((Result<ASAuthorization, Error>) -> Void)
)

- label:
    - 이 객체에서 제공하는 레이블 중 하나를 사용할 수 있습니다.
    - 종류: .signIn, .continue, .signUp (default가 signIn이라 생략 가능하지만 이해가 쉽도록 명시했습니다)
- onRequest:
    - 버튼이 탭 되면 호출되며, Sign In with Apple을 위한 request가 만들어집니다.
    - 이 클로저 안에서 request를 커스터마이징할 수 있습니다. (.fullName, email 등 요청하고자 하는 정보)
- onCompletion:
    - 위에서 보낸 request의 result를 받아 처리할 내용을 작성할 수 있습니다.

minsson committed:
#9 feat: randomNonceString 메서드 구현

- Firebase의 Apple 계정으로 로그인 공식문서 참고
- 관련 링크: https://firebase.google.com/docs/auth/ios/apple?authuser=0&hl=ko
- 설명:
    - 모든 로그인 요청에 대해 임의의 문자열("nonce")을 생성하여 앱의 인증 요청에 대한 응답으로 받은 ID 토큰이 특별히 부여되었는지 확인하는 데 사용할 수 있습니다.
    - 이 단계는 replay attacks을 방지하는 데 중요합니다.
    - SecRandomCopyBytes(_:_:_) 를 사용하여 암호화된 보안 nonce를 생성할 수 있습니다.

minsson committed:
#9 feat: sha256 메서드 구현

- Firebase의 Apple 계정으로 로그인 공식문서 참고
- 관련 링크: https://firebase.google.com/docs/auth/ios/apple?authuser=0&hl=ko
- 설명:
    - 로그인 요청과 함께 nonce의 SHA256 해시를 보내면 Apple이 응답에서 변경하지 않고 전달합니다.
    - Firebase는 원래 nonce를 해싱하고 이를 Apple에서 전달한 값과 비교하여 응답의 유효성을 검사합니다.
- Firebase에서는 @available(iOS 13, *)을 메서드 앞에 붙여주었지만, 우리 타겟은 원래 그보다 높으므로 생략했습니다.

minsson committed:
#9 feat: handleSignInWithAppleRequest 메서드 구현 및 적용

- Firebase의 Apple 계정으로 로그인 공식문서 참고
- 관련 링크: https://firebase.google.com/docs/auth/ios/apple?authuser=0&hl=ko
- 설명:
    - SignInWithAppleButton가 탭 되었을 때 request를 설정하기 위한 메서드입니다.
    - request에 nonce를 더해줍니다.
        - nonce는 replay attacks를 막기 위한 랜덤 넘버입니다.
    - nonce를 만들기 위해 randomNonceString()를 사용했습니다.
    - 이 nonce를 sha256으로 암호화 해 request.nonce에 넣었습니다.
    - SignInWithAppleButton 객체는 이 request를 받아 ASAuthorizationAppleIDProvider에 전달합니다.
        - signIn flow를 진행하기 위함입니다.

minsson committed:
#9 feat: handleSignInWithAppleCompletion 메서드 임시 구현 및 적용

- Firebase의 Apple 계정으로 로그인 공식문서 참고
- 관련 링크: https://firebase.google.com/docs/auth/ios/apple?authuser=0&hl=ko
- 설명:
    - SignInWithAppleButton를 통해 sign-in 과정이 완료되었을 때 호출되는 메서드입니다.
    - Result 타입의 결과를 받아 성공 여부에 따른 로직 분기 처리를 진행합니다.
    - 임시 구현 상태이므로 로직이 완성되지 않은 상태입니다.

minsson committed:
#9 feat: SignInWithApple이 성공하면 FirebaseAuth의 credential 생성

- Firebase의 Apple 계정으로 로그인 공식문서 참고
- 관련 링크: https://firebase.google.com/docs/auth/ios/apple?authuser=0&hl=ko
- 설명:
    - Sign-In Flow의 결과가 success인 경우, 받은 데이터의 credential 속성에서 필요한 정보 추출
    - FirebaseAuth에 필요한 nonce, appleIDToken의 인코딩된 값을 추출하기 위한 과정 진행
    - 취합된 정보를 바탕으로 FirebaseAuth에서 사용하기 위한 OAuthProvider.credential 생성

minsson committed:
#9 feat: firebaseAuthCredential로 FirebaseAuth에 로그인하는 기능 구현

- Firebase의 Apple 계정으로 로그인 공식문서 참고
- 관련 링크: https://firebase.google.com/docs/auth/ios/apple?authuser=0&hl=ko

minsson committed:
#9 refactor: 보안 관련 코드 AuthSecurityService로 분리

- 뷰 모델의 역할이 과중하여 별도의 객체로 분리했습니다.

minsson committed:
#9 refactor: AppleLoginManager 객체 생성 및 기존 뷰모델의 역할 이양

- 뷰 모델의 역할이 과중하여 별도의 객체로 분리했습니다.
- 기존 코드에서는 iOS14부터 SwiftUI에서 이용 가능한 SignInWithAppleButton 객체를 사용했습니다.
- 처음에는 구현이 매우 간단해 사용했지만, 뷰에서 비즈니스 로직을 실행하고, 뷰가 뷰모델에게 데이터를 전달하게 되는 문제가 있었습니다.
- 더 좋은 구조 분리를 위해 코드 구현이 번거롭지만 FirebaseAuth에서 안내하는 기존의 방식으로 리팩토링하기 시작했습니다.

minsson committed:
#9 refactor: startSignInWithAppleFlow 메서드 역할 분리

minsson committed:
#9 refactor: AppleLoginManager의 nonce에 기본값 부여

- AppleLoginManager 내부에서 전역적으로 사용되므로 기본값을 부여해 불필요한 코드를 제거했습니다.

minsson committed:
#9 refactor: FirebaseAuth에 넘길 정보를 추상화한 모델 객체 생성

- 기존 코드에서는 AppleLoginManager에서 FirebaseAuth 로그인 과정까지 구현했습니다.
- FirebaseAuth를 별도의 객체로 분리하기 위해, AppleLoginManager 객체가 넘겨줘야할 정보를 추상화했습니다.

minsson committed:
#9 refactor: FirebaseLoginManager 타입 생성

- AppleLogin 등 소셜 로그인으로부터 토큰을 받고, FirebaseAuth를 통해 Firebase에 로그인 시키는 역할의 객체입니다.

minsson committed:
#9 refactor: 애플 로그인 토큰 정보를 컴플리션 핸들러로 전달하기 위한 기능 구현

- signIn(completion:) 메서드는 FirebaseLoginManager에서 실행해 애플 로그인 토큰 등 크레덴셜을 전달해주기 위한 메서드입니다.
    - 컴플리션 핸들러를 통해 전달합니다.
- AppleLoginManager 객체 내에 signInCompletion 프로퍼티를 선언 후, signIn(completion:)의 completion을 할당했습니다.
    - ASAuthorizationControllerDelegate의 didCompleteWithAuthorization 메서드 실행 시점을 잡기 위함입니다.
- 클로저는 참조타입이므로 didCompleteWithAuthorization에서 signInCompletion에 결과값을 전달하면, completion에 전달한 것과 같아집니다.

minsson committed:
#9 refactor: 코드 들여쓰기 수정

- Firebase 공식문서에서 가져와 들여쓰기 깊이가 다른 부분을 수정했습니다.

minsson committed:
#9 refactor: FirebaseLoginManager의 signIn 메서드 구현

- AppleLoginManager로부터 FirebaseAuth에 필요한 데이터를 컴플리션 핸들러로 받아와, FirebaseAuth를 진행하는 메서드입니다.
    - 이 메서드는 LoginViewModel에서 로그인 버튼이 탭 되면 호출할 메서드입니다.
- Firebase 로그인이 완료되면 컴플리션 핸들러로 성공 여부를 반환합니다.
- 소셜 로그인 관련 확장성을 대비하여 AppleLoginManager는 추후 SocialLoginManager로 추상화하려고 합니다.
    - SocialLoginManager 프로토콜을 채택한 객체라면 어떤 서비스든 FirebaseAuth로 Firebase 로그인을 할 수 있게 됩니다.

minsson committed:
#9 refactor: AuthenticationViewModel을 LoginViewModel로 네이밍 변경

- LoginView와도 연결되면서, 더 쉽고 직관적인 네이밍을 쓰고자 변경했습니다.

minsson committed:
#9 refactor: 리팩토링에 따라 불필요해진 LoginViewModel 기존 코드 삭제

minsson committed:
#9 refactor: UIKit의 애플로그인 버튼을 래핑하여 애플로그인 버튼 생성 및 적용

- 기존 SwiftUI API에서 제공하는 SignInWithAppleButton의 경우 필연적으로 뷰에 로직이 포함되는 구조였습니다.
- 뷰와 비즈니스 로직을 완전히 분리하고자 UIKit에서 사용하는 ASAuthorizationAppleIDButton을 사용해 로그인 버튼을 구현했습니다.

minsson committed:
#9 feat: Sign in with Apple Capability 추가

minsson committed:
#9 refactor: 애플로그인 버튼을 탭하면 뷰모델의 메서드 호출 (뷰에서 로직 제거 완료)

- 뷰와 뷰모델의 로직을 분리하기 위한 리팩토링이 완료되었습니다.

minsson committed:
#9 refactor: AppleLoginButtonView 접근제어자 및 파일 위치 수정

- LoginView에서만 사용될 예정이므로 fileprivate 키워드를 붙이고 LoginView 내부로 이동했습니다.

minsson committed:
#9 feat: AppleLoginButtonView가 다크모드에 대응하는 기능 추가

minsson committed:
#9 feat: FirebaseAuth 자동로그인 기능 구현

- LoginVIewModel의 init 시점에 authManager(FirebaseLoginManager)를 통해 기존에 로그인 되어 있었는지 확인합니다.
    - FirebaseLoginManager 객체의 shouldSignIn을 통해 현재 애플 아이디를 통한 기존 로그인 여부를 확인합니다.
        - 결과적으로 새로 로그인해야 한다면 true, 로그인하지 않아도 된다면 false를 반환합니다.
- 위의 결과는 LoginViewModel의 shouldSignIn에 할당되며, 이를 구독하고 있던 LoginView에서 어떤 뷰를 보여줄지 결정합니다.

minsson committed:
#9 refactor: AppleLoginManager를 SocialLoginManagerInterface로 추상화

- 추후 애플 로그인 외에도 다른 로그인 기능이 추가될 가능성이 있어 추상화 진행 및 적용했습니다.
- 테스트 용이성도 고려했습니다.

minsson committed:
#9 refactor: FirebaseLoginManager를 ServerLoginManagerInterface로 추상화

- 테스트 용이성을 고려했습니다.
- 별도의 서버를 운용하지 않는 이상 Firebase를 계속 사용할 예정이라 확장 가능성은 유의미하지 않다고 생각합니다.

minsson committed:
#9 refactor: LoginView가 LoginViewModel을 주입받도록 수정

minsson committed:
#9 refactor: authManager 네이밍을 serverLoginManager로 변경

minsson committed:
#9 refactor: Protocol의 네이밍을 Interface 대신 Protocol로 변경

minsson committed:
#9 refactor: LoginViewModel의 serverLoginManager 기본값 제거

- 기본값 없이 완전히 의존성 주입을 받기 위함입니다.

minsson committed:
#9 refactor: LoginViewModel이 소셜로그인매니저에 의존하지 않도록 수정

- LoginViewModel의 관심사는 로그인을 시작하고, 완료 여부를 반환 받는 것입니다.
- 탭 된 소셜로그인 버튼의 종류를 serverLoginManager에 전달해줄 필요는 있지만, 어떤 구체 소셜로그인매니저인지는 알 필요가 없습니다.
- 따라서 SocialLogin이라는 enum 타입을 만들고, 해당 타입의 case만 알면 각 구체소셜로그인매니저 타입의 인스턴스가 반환되도록 수정했습니다.

minsson committed:
#9 chore: 파일 및 폴더 정리

minsson committed:
#9 feat: LoginView에 앱 로고 추가

- 아직 디자인시스템에 앱 로고를 위한 폰트가 추가되어있지 않으므로, 우선은 가장 비슷한 느낌의 폰트를 골랐습니다.
- 디자인시스템 구현 완료 후 앱 로고에 폰트 적용이 필요합니다.

minsson committed:
#9 refactor: 로그인 완료 후 TabBarContainerView가 보이도록 수정

- 탭 바와 GalpiGridView 모두를 포함하는 TabBarContainerView가 보이도록 했습니다.

minsson committed:
#9 fix: 애플 로그인을 완료해도 로그인이 되지 않던 문제 해결

- AppleLoginManager 객체는 ASAuthorizationController를 통해 애플 로그인 request를 보냅니다.
- request에 대한 응답은 ASAuthorizationControllerDelegate의 델리게이트 메서드를 통해 받습니다.
- 위 과정에서 AppleLoginManager 객체가 델리게이트 메서드가 실행되기도 전에 메모리에서 해제되는 것이 문제였습니다.
- 이에 따라 FirebaseLoginManager가 socialLoginManager 타입을 알도록 바꾸었습니다.

minsson committed:
#9 refactor: SocialLoginCredential이 providerID 대신 socialLoginName을 받도록 수정

- 기존에는 AppleLoginManager 등 각 SocialLoginManagerProtocol의 구체 타입이 FirebaseAuth에 쓰일 String을 넘겨준다는 점이 아쉬웠습니다.
- SocialLoginManagerProtocol의 구체 타입에서 FirebaseAuth를 쓰는 모습을 보이지 않아야 역할이 더 명확하게 분리되었다고 생각했기 때문입니다.
- 따라서 SocialLoginCredential의 이니셜라이저로 providerID 대신 socialLoginName으로 SocialLogin 타입을 받아, FirebaseLoginManager 쪽에서 socialLoginName.firebaseProviderID와 같은 방식으로 정보를 꺼내 쓸 수 있도록 수정했습니다.

minsson commented on galpi/galpi/Services/OAuth/SocialLogin/SocialLogin.swift:
```diff
@@ -0,0 +1,26 @@
+        }
+    }
+    
+    var firebaseProviderID: String {
+        switch self {
+        case .apple:
+            return "apple.com"
+        }
+    }
```
f019f64926cdb66b1e1f9bb6a7b05f7f3196feda 커밋 메시지에 이와 관련된 내용이 있습니다.

minsson replied:
이 코멘트는 처음 PR 보내자마자 작성했던 코멘트인데, 그때 코멘트 최종 제출을 안 해서 이제야 나오네요..! 이 코멘트도 확인 부탁드려요.

minsson replied:
이 코멘트는 참고용으로 처음 PR을 보낼 때 작성했던 코멘트인데, 그때 코멘트 최종 제출을 안 해서 이제야 나오네요..! 이 코멘트도 확인 부탁드려요. ☺️

minsson commented on galpi/galpi/Services/OAuth/FirebaseLoginManager.swift:
```diff
@@ -0,0 +1,58 @@
+        return Auth.auth().currentUser != nil ? false : true
+    }
+    
+    var socialLoginManager: SocialLoginManagerProtocol?
```
174b88bc6e947940955caec90a823ee33842fab7 의 커밋메시지 내용으로 인해 FirebaseLoginManager가 SocialLoginManagerProtocol을 들고 있게 되었습니다.

minsson replied:
이 코멘트는 참고용으로 처음 PR을 보낼 때 작성했던 코멘트인데, 그때 코멘트 최종 제출을 안 해서 이제야 나오네요..! 이 코멘트도 확인 부탁드려요. ☺️

jcrescent61 commented on galpi/galpi/Domain/Login/LoginViewModel.swift:
```diff
@@ -0,0 +1,37 @@
+    }
+    
+    private func executeServerLogin(with socialLogin: SocialLogin) {
+        serverLoginManager.signIn(with: socialLogin) { boolResult in
```
P1: boolResult 보다 다른 네이밍이 좋을 것 같아요.
isSignIn과 같이 어떤 행위에대한 불리언인지 짐작할 수 있는 네이밍이 좋을 것 같습니다.

minsson replied:
test

minsson replied:
엘렌이 말씀하신 부분은 수정을 거쳐 현재 아래와 같습니다.

```swift
    private func executeServerLogin(with socialLogin: SocialLogin) {
        serverAuthService.signIn(with: socialLogin) { boolResult in
            switch boolResult {
            case .success(let isSignedIn):
                self.isSignedIn = isSignedIn
            case .failure(let error):
                print(error)
            }
        }
    }
```

저도 처음에는 isSignedIn을 생각했는데, 위 코드에서 boolResult는 `Result<Bool, FirebaseAuthError>` 타입이라 isSignedIn을 쓰면 그냥 Bool 타입처럼 보일 수 있어 직관적인 이름을 고민하다가 boolResult라는 네이밍을 선택했었습니다.

혹시 더 직관적인 네이밍이 있을까요..!?

jcrescent61 replied:
차라리 boolResult보다는 result가 더 직관적일 수 있겠다는 생각이 드네요..! .success에서 isSignedIn을 설명해주고있기 때문입니다!

jcrescent61 commented on galpi/galpi/Services/OAuth/FirebaseLoginManager.swift:
```diff
@@ -0,0 +1,58 @@
+final class FirebaseLoginManager: ServerLoginManagerProtocol {
+    
+    var shouldSignIn: Bool {
+        return Auth.auth().currentUser != nil ? false : true
```
P3: 개인적으로 가독성을 위해서 `!=`을 지양하는 편입니다. `==`으로 조건을 만들어서 한번 더 생각할 거리를 줄여주는것을 선호합니다!

minsson replied:
저도 공감합니다! 
우선 부등호를 등호로 바꾸고, 이와 관련하여 변수명도 수정해 더 자연스럽게 읽히도록 바꾸었습니다.

### 관련 커밋
1) e203cdc55a5375c81ce99410b0338a01e87a701c
[#9](https://github.com/team-galpi/ios-galpi/issues/9) refactor: 가독성 개선을 위해 비교연산자 사용시 부등호 대신 등호를 사용하도록 수정
- 부등호로 인해 코드를 읽을 때 한번 더 생각해야 했습니다. 좀 더 자연스럽게 읽히도록 수정했습니다.

2) e2a2bbbbe685a3ce104b797bf6838427fb75b4fc
[#9](https://github.com/team-galpi/ios-galpi/issues/9) refactor: 로그인 과정의 shouldSignIn을 isSignedIn으로 변경
- Bool 타입으로 로그인 성공 여부를 전달하는데, 기존의 네이밍에서는 성공하면 false를 전달하게 돼 사고의 흐름이 자연스럽지 않았습니다.
- 특히 FirebaseLoginManager의 signIn 메서드에서 completion을 통해 true를 넘겨주고, LoginViewModel의 isSignedIn에 그 값을 넘겨주는 것이 좀 더 자연스러워졌습니다.

jcrescent61 commented on galpi/galpi/Services/OAuth/AuthSecurityService.swift:
```diff
@@ -0,0 +1,57 @@
+import Foundation
+import CryptoKit
+
+struct AuthSecurityService {
```
P3: AuthSecurityService 객체일까요? NonceManager라는 네이밍이 더 적합하지 않을까요?

minsson replied:
제가 보기에도 NonceManager라는 네이밍이 기존의 네이밍보다 더 좋아 보이네요! ☺️

특히 Service와 Manager의 역할에 대해 다시 한번 생각해보고, 검색해봤는데, 여기서는 Manager라는 이름이 더 적절해보였어요. 

다만 Nonce와 AuthSecurity는 아직 확신이 안 서네요. 가령 sha256이 nonce만을 위한 private func라면 확실히 Nonce가 될 것 같은데, nonce 관련해서 쓰는 거긴 하지만, sha256 메서드도 별도로 호출되고 있어서 네이밍이 모호하게 느껴져요.

우선은 NonceManager로 수정했는데, AuthSecurityManager와 같은 네이밍에 대해서는 어떻게 생각하시나요?

관련 커밋: b91ab3f96b7180b82e0a90a8a213949f88feb97d

jcrescent61 reviewed (commented):
민쏜! 고생많으셨습니다! 어찌저찌 늦게 답변을 달게되었네요! 죄송합니다🙇🏻‍♀️
제가 봤을때 크게 이상한 점은 없었던 것 같습니다. 개인적인 스타일이 달라서 코멘트 몇개 달아보았습니다.
리뷰 반영하시고 다시 DM 주세요!

minsson committed:
#9 refactor: if let 바인딩 컨벤션에 맞게 수정 및 들여쓰기 수정

- 수정 전 코드가 더 편한 방법이지만, 아직 다른 부분에 기존의 방법으로 작성한 코드가 있어 우선 컨벤션에 맞게 수정했습니다.

minsson committed:
#9 refactor: 가독성 개선을 위해 비교연산자 사용시 부등호 대신 등호를 사용하도록 수정

- 부등호로 인해 코드를 읽을 때 한번 더 생각해야 했습니다. 좀 더 자연스럽게 읽히도록 수정했습니다.

minsson committed:
#9 refactor: 로그인 과정의 shouldSignIn을 isSignedIn으로 변경

- Bool 타입으로 로그인 성공 여부를 전달하는데, 기존의 네이밍에서는 성공하면 false를 전달하게 돼 사고의 흐름이 자연스럽지 않았습니다.
- 특히 FirebaseLoginManager의 signIn 메서드에서 completion을 통해 true를 넘겨주고, LoginViewModel의 isSignedIn에 그 값을 넘겨주는 것이 좀 더 자연스러워졌습니다.

minsson committed:
#9 refactor: Result 타입으로 전달 받은 Bool 타입 결과를 할당하도록 수정

- 기존에는 왜 self.isSignedIn에 true가 들어가야 하는지 이해하기 어려웠습니다.
- 가독성 개선을 위해 Result 타입으로 전달 받은 내용이 success인 경우를 isSignedIn이라고 네이밍하고, 그 값을 self.isSignedIn에 할당했습니다.

minsson committed:
#9 refactor: AuthSecurityService 네이밍 NonceManager로 변경

minsson committed:
#9 refactor: ServerLoginManagerProtocol의 네이밍 변경

- ServerAuthServiceProtocol로 변경

minsson committed:
#9 refactor: FirebaseLoginManager 네이밍 변경

- 여러 매니저를 활용해 하나의 큰 로직을 수행하므로 FirebaseAuthService로 변경했습니다.

minsson committed:
#9 refactor: serverLoginManager 변수 네이밍 변경

- FirebaseLoginManager 타입이 FirebaseAuthService로 변경되었으므로, 이에 맞춰 관련 변수명을 수정했습니다.

minsson commented:
엘렌, 코멘트 달아주신 부분에 대한 수정은 우선 완료되었습니다. ☺️

minsson reviewed (commented):
엘렌, 말씀하신 부분에 대해 코멘트 달아두었습니다!
확인 부탁드려요 ☺️

minsson commented:
엘렌, 코드리뷰에 코멘트 달았던 게 전화로 말씀드렸던 것처럼 좀 꼬여서 다시 코멘트를 남겼습니다! 확인 부탁드려요 ☺️

그리고 에러 처리는 별도 PR로 빼서 공용으로 쓰일 수 있도록 만들어보려고 합니다!
따라서 이번 PR에서 에러 처리는 모두 `print(error)`로만 임시 구현해두었습니다. 참고 바랍니다! 😁

minsson reviewed (commented):
엘렌, 코드리뷰에 코멘트 달았던 게 전화로 말씀드렸던 것처럼 좀 꼬여서 다시 코멘트를 남겼습니다! 확인 부탁드려요 ☺️

그리고 에러 처리는 별도 PR로 빼서 공용으로 쓰일 수 있도록 만들어보려고 합니다!
따라서 이번 PR에서 에러 처리는 모두 print(error)로만 임시 구현해두었습니다. 참고 바랍니다! 😁
